---
title: "host-microbiome"
author: "John McGInniss"
date: '2022-03-15'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Background

I'm interested in host-microbe interactions in the lung and how they influence lung health. Recently, our group performed a study looking at outcomes after lung transplant in relation to the lung  microbiome and inflammatory markers. A challenge was how to compare these two datasets.

The samples are from the lung in the hour after lung transplantation. The microbiome data is from 16S amplicon sequencing and the inflammatory data is from a 41-plex Luminex assay. So for each sample we have both data. My initial analysis basically looked at both separately and how they related to health and disease post-transplant and then correlations between microbial taxa and the inflammatory mediators. 

Inspired by my teams' work on agent based modeling, I wanted to take a modeling approach to the data which I think moves this in a more mechanistic direction. 

First:
1. Take a look at the data
2. Simulate models with DSAIRM looking at bacterial-immune interactions
3. Compare actual models to simulated to understand possible mechanistic link -- which best aligns? (open to suggestions here...)


## Load Packages
```{r packages}
library(tidyverse)
library(tidymodels)  # for the parsnip package, along with the rest of tidymodels

# Helper packages
library(readr)       # for importing data
library(broom.mixed) # for converting bayesian models to tidy tibbles
library(dotwhisker)  # for visualizing regression results
library(rstanarm)
library(skimr)

# Main package of interest
library('DSAIRM')
```

## Data

```{r data}
# My data
microbiome_cytokine <- 
  read_tsv("data/microbiome_cytokine.tsv")

#DF with sample, top taxa and inflammatory markers

glimpse(microbiome_cytokine)

# Distribution of cytokines
microbiome_cytokine %>% 
  pivot_longer(cols = c(1:42), names_to = "variable", values_to = "value") %>% 
  group_by(variable) %>% 
  summarise(median = median(value), iqr25 = quantile(value, 0.25), iqr75 = quantile(value, 0.75)) %>% 
  ggplot(aes(x=reorder(variable, -median), y=median)) + 
  geom_point() + 
  geom_linerange(aes(ymin=iqr25, ymax=iqr75)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

# Distribution of taxa
microbiome_cytokine %>% 
  pivot_longer(cols = c(43:58), names_to = "variable", values_to = "value") %>% 
  group_by(variable) %>% 
  summarise(median = median(value), iqr25 = quantile(value, 0.25), iqr75 = quantile(value, 0.75)) %>% 
  ungroup() %>% 
  ggplot(aes(x=reorder(variable, -median), y=median)) + 
  geom_point() + 
  geom_linerange(aes(ymin=iqr25, ymax=iqr75)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

# Correlation plot
cor_df <- cor(scale(microbiome_cytokine), method = "spearman")
corrplot::corrplot(cor_df)

cor_df[1:41,43:58] > abs(0.2)

```

Overall there are modest correlations between taxa and cytokines. Can a modeling approach help look at possible mechanistic links?

Will turn to Dynamical Systems Approaches to Immune Response Modeling (DSAIRM).
1. First will look at the GUI and examine model
2. Then will tune some model parameters
(3. Can also play with/adjust the underlying ODE compartmental models)

# DSAIRM
```{r}

dsairmmenu()
help('simulate_basicbacteria_ode')

# Bacterial-immune default model
result <- simulate_basicbacteria_ode(
  B = 100,      # starting value bacteria
  I = 1,        # starting value for immune response
  g = 1,        # maximum rate of bacterial growth
  Bmax = 1e+05, # bacterial carrying capacity
  dB = 0.5,     # bacterial death rate
  k = 1e-04,    # rate of bacteria killed by immune response
  r = 1e-04,    # immune response growth rate
  dI = 2,       # immune response decay rate
  tstart = 0,   # start time of simulation
  tfinal = 100, # final time of simulation
  dt = 0.01     # times which result returned
)

result_plot <- result$ts

result_plot <- result_plot %>% 
  pivot_longer(cols = c(2:3), names_to = "Agent", values_to = "Amount")

ggplot(result_plot, aes(x=time, y=Amount, color=Agent)) + geom_point()
```

## Modify the immune response for a given bacterial load

```{r}
# Bacterial-immune low activation model
result_low <- simulate_basicbacteria_ode(
  B = 100,      # starting value bacteria
  I = 1,        # starting value for immune response
  g = 1,        # maximum rate of bacterial growth
  Bmax = 1e+05, # bacterial carrying capacity
  dB = 0.5,     # bacterial death rate
  k = 1e-04,    # rate of bacteria killed by immune response
  r = 1e-05,    # low immune response growth rate
  dI = 2,       # immune response decay rate
  tstart = 0,   # start time of simulation
  tfinal = 100, # final time of simulation
  dt = 0.01     # times which result returned
)

result_low_plot <- result_low$ts

result_low_plot <- result_low_plot %>% 
  pivot_longer(cols = c(2:3), names_to = "Agent", values_to = "Amount") %>% 
  ggplot(., aes(x=time, y=Amount, color=Agent)) + 
  geom_point()

# Bacterial-immune mid activation model
result_mid <- simulate_basicbacteria_ode(
  B = 100,      # starting value bacteria
  I = 1,        # starting value for immune response
  g = 1,        # maximum rate of bacterial growth
  Bmax = 1e+05, # bacterial carrying capacity
  dB = 0.5,     # bacterial death rate
  k = 1e-04,    # rate of bacteria killed by immune response
  r = 3e-04,    # low immune response growth rate
  dI = 2,       # immune response decay rate
  tstart = 0,   # start time of simulation
  tfinal = 100, # final time of simulation
  dt = 0.01     # times which result returned
)

result_mid_plot <- result_mid$ts

result_mid_plot <- result_mid_plot %>% 
  pivot_longer(cols = c(2:3), names_to = "Agent", values_to = "Amount") %>% 
  ggplot(., aes(x=time, y=Amount, color=Agent)) + 
  geom_point()

# Bacterial-immune high activation model
result_high <- simulate_basicbacteria_ode(
  B = 100,      # starting value bacteria
  I = 1,        # starting value for immune response
  g = 1,        # maximum rate of bacterial growth
  Bmax = 1e+05, # bacterial carrying capacity
  dB = 0.5,     # bacterial death rate
  k = 1e-04,    # rate of bacteria killed by immune response
  r = 1e-03,    # low immune response growth rate
  dI = 2,       # immune response decay rate
  tstart = 0,   # start time of simulation
  tfinal = 100, # final time of simulation
  dt = 0.01     # times which result returned
)

result_high_plot <- result_high$ts

result_high_plot <- result_high_plot %>% 
  pivot_longer(cols = c(2:3), names_to = "Agent", values_to = "Amount") %>% 
  ggplot(., aes(x=time, y=Amount, color=Agent)) + 
  geom_point()
```

# Supply a vector for immune activation rate, capture max B and I
```{r}
#values for immune activation rate, r, for which to run the simulation
rvec <- 10^seq(-5,-2,length=20)  

#this variable will hold the peak values for each r
Bpeak <- rep(0,length(rvec))
Ipeak <- rep(0,length(rvec))
for (n in 1:length(rvec))   
{
  #run the simulation model for different values of r 
  #for simplicity, all other inputs are kept at their defaults
  result <- simulate_basicbacteria_ode(r = rvec[n])
  #record max number of bacteria for each value of r
  Bpeak[n] <- max(result$ts[,"B"])
  Ipeak[n] <- max(result$ts[,"I"])
}
#plot final result
rvec_sim <- cbind(rvec, Ipeak, Bpeak)

plot(Bpeak, Ipeak)
plot(Bpeak/5000, log10(Ipeak))

# Make some adjustments to approximate data type
rvec_sim_adj <- cbind(rvec, Ipeak_log10 = log10(Ipeak+1e-5), Bpeak_rel_abd = (Bpeak/50000))

```

# Fit linear models to the data suggested by correlations
```{r}
Vars <- as.list(colnames(microbiome_cytokine[,43:58])) # Create list of taxa

Outcomes <- as.list(colnames(microbiome_cytokine[,1:41])) # List of cytokines

allModelsList <- lapply(paste(Outcomes, " ~ ", Vars), as.formula)

allModelsResults <- lapply(allModelsList, function(x) lm(x, data = microbiome_cytokine))

allModelsSummaries <- lapply(allModelsResults, summary)

for (i in Outcomes){
  lapply(paste(Outcomes[i], " ~ ", Vars), as.formula)
}
```


## Fit curve to the simulated data

```{r}
#fit polynomial regression models up to degree 5
fit1 <- lm(Ipeak_log10~Bpeak_rel_abd, data=df)
fit2 <- lm(Ipeak_log10~poly(Bpeak_rel_abd,2,raw=TRUE), data=rvec_sim_adj)
fit3 <- lm(Ipeak_log10~poly(Bpeak_rel_abd,3,raw=TRUE), data=rvec_sim_adj)
fit4 <- lm(Ipeak_log10~poly(Bpeak_rel_abd,4,raw=TRUE), data=rvec_sim_adj)
fit5 <- lm(Ipeak_log10~poly(Bpeak_rel_abd,5,raw=TRUE), data=rvec_sim_adj)

#create a scatterplot of x vs. y
plot(rvec_sim_adj$Bpeak_rel_abd, rvec_sim_adj$Ipeak_log10, pch=19, xlab='Bacterial burden', ylab='Peak immune response')

#define x-axis values
x_axis <- seq(1, 15, length=15)

#add curve of each model to plot
lines(x_axis, predict(fit1, data.frame(x=x_axis)), col='green')
lines(x_axis, predict(fit2, data.frame(x=x_axis)), col='red')
lines(x_axis, predict(fit3, data.frame(x=x_axis)), col='purple')
lines(x_axis, predict(fit4, data.frame(x=x_axis)), col='blue')
lines(x_axis, predict(fit5, data.frame(x=x_axis)), col='orange')
```


## Compare to observed data


# Bayesian modeling
```{r}
# Simulate scenario of positive correlation, negative, neutral

microbiome_cytokine %>% 
  select(43:58) %>% # select the taxa
  pivot_longer(cols = c(1:16), names_to = "Taxa", values_to = "Rel_abd") %>% 
  ggplot(., aes(Rel_abd)) + geom_histogram()
  

x1 <- rnbinom(111, mu = 5, 1) / 100 # Generate independent variable based on rel abd
x1 <- x1 %>% as_tibble_col(column_name = "Rel_abd")
hist(x1$Rel_abd)
error <- rnorm(111,0,50) / 100
hist(error)

#Generate the dependent variable
microbiome_cytokine %>% 
  select(1:41) %>% # select the taxa
  pivot_longer(cols = c(1:41), names_to = "Cytokine", values_to = "Log10") %>% 
  summarize(mean = mean(Log10), sd = sd(Log10))

# Simulated models of positive, neg, and neutral relationship.
y_positive <- 3.9 + (5*x1) + error
y_positive <- y_positive %>% as_tibble_col(column_name = "Log10_pos")

y_negative <- 3.9 - (5*x1) + error
y_negative <- y_negative %>% as_tibble_col(column_name = "Log10_neg")

y_neutral <- 3.9 - (0*x1) + error
y_neutral <- y_neutral %>% as_tibble_col(column_name = "Log10_neut")

# Put data together
simulated_data <- cbind(y_positive$Log10_pos,  y_neutral$Log10_neut, y_negative$Log10_neg, x1$Rel_abd)

colnames(simulated_data) <- c("Log10_pos", "Log10_neut", "Log10_neg", "Rel_abd")

#Create the simulated models
# Positive model
bayes_positive <- stan_lm(Log10_pos ~ Rel_abd,
        data = simulated_data,
        prior = R2(what = "mean", location = 0.5),
        seed=123)

summary(bayes_positive)

# Negative model
bayes_negative <- stan_lm(Log10_neg ~ Rel_abd,
        data = simulated_data,
        prior = R2(what = "mean", location = 0.5),
        seed=123)

summary(bayes_negative)

# Neutral model
bayes_neutral <- stan_lm(Log10_neut ~ Rel_abd,
        data = simulated_data,
        prior = R2(what = "mean", location = 0.5),
        seed=123)

summary(bayes_neutral)

```

# Model each variable
```{r}
# train the model for the rest of variables
bayes_1 <- stan_lm(IL1a ~ Prevotella,
        data = microbiome_cytokine,
        prior = R2(what = "mean", location = 0.5),
        seed=123)


## Function for rest of variables ##

```


```{r}
# Compare models with LOO
loo_1 <- loo(bayes_1, cores = 2)
loo_negative <- loo(bayes_negative, cores = 2)
loo_positive <- loo(bayes_positive, cores = 2)
loo_neutral <- loo(bayes_neutral, cores = 2)

loo_compare(loo_1, loo_negative, loo_positive, loo_neutral)


```




